File: add_to_app/multiple_flutters_ios/MultipleFluttersIos/DataModel.swift
// Copyright 2021 The Flutter team. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import Foundation

/// A protocol that receives updates when the datamodel is changed.
protocol DataModelObserver: AnyObject {
  func onCountUpdate(newCount: Int64)
}

/// A wrapper object around a weak reference to an object that implements DataModelObserver.
///
/// This is required since you can't directly hold weak references to protocols in data structures.
private struct AnyDataModelObserver {
  weak var base: DataModelObserver?

  init(_ base: DataModelObserver) {
    self.base = base
  }
}

/// A singleton data model that is observable.
class DataModel {
  private var _count: Int64 = 0
  var count: Int64 {
    get {
      return self._count
    }
    set {
      self._count = newValue
      for observer in observers {
        if let base = observer.base {
          base.onCountUpdate(newCount: self._count)
        }
      }
    }
  }
  private var observers: [AnyDataModelObserver] = []
  static let shared = DataModel()

  func addObserver(observer: DataModelObserver) {
    observers.append(AnyDataModelObserver(observer))
  }

  func removeObserver(observer: DataModelObserver) {
    observers.removeAll { (element: AnyDataModelObserver) -> Bool in
      if let base = element.base {
        return base === observer
      } else {
        return true
      }
    }
  }
}

File: add_to_app/multiple_flutters/multiple_flutters_android/app/src/main/java/dev/flutter/multipleflutters/DataModel.kt
package dev.flutter.multipleflutters

import java.lang.ref.WeakReference

/**
 * Interface for getting notifications when the DataModel is updated.
 */
interface DataModelObserver {
    fun onCountUpdate(newCount: Int)
}

/**
 * A singleton/observable data model for the data shared between Flutter and the host platform.
 *
 * This is the definitive source of truth for all data.
 */
class DataModel {
    companion object {
        val instance = DataModel()
    }

    private val observers = mutableListOf<WeakReference<DataModelObserver>>()

    public var counter = 0
        set(value) {
            field = value
            for (observer in observers) {
                observer.get()?.onCountUpdate(value)
            }
        }

    fun addObserver(observer: DataModelObserver) {
        observers.add(WeakReference(observer))
    }

    fun removeObserver(observer: DataModelObserver) {
        observers.removeIf {
            if (it.get() != null) it.get() == observer else true
        }
    }
}
